source/MaterialXGenGlsl/GlslShaderGenerator.cpp

```diff
@@ -266,492 +237,331 @@ GlslShaderGenerator::GlslShaderGenerator()
-        {
-            const string& type = _syntax->getTypeName(uniform->type);
-            shader.addLine(type + " " + uniform->name);
-        }
-        shader.endScope(true);
-        shader.newLine();
-        shader.addLine("uniform " + lightData.name + " " + lightData.instance + "[MAX_LIGHT_SOURCES]");
-        shader.newLine();
+        const VariableBlock& lightData = stage.getUniformBlock(HW::LIGHT_DATA);
+        emitLine("struct " + lightData.getName(), stage, false);
+        emitScopeBegin(stage, ShaderStage::Brackets::BRACES);
+        emitVariableDeclarations(lightData, EMPTY_STRING, SEMICOLON, context, stage, false);
+        emitScopeEnd(stage, true);
+        emitLineBreak(stage);
+        emitLine("uniform " + lightData.getName() + " " + lightData.getInstance() + "[MAX_LIGHT_SOURCES]", stage);
+        emitLineBreak(stage);
     }
 
-    // Add vertex data block
-    if (!vertexDataBlock.empty())
+    // Add vertex data inputs block
+    const VariableBlock& vertexData = stage.getInputBlock(HW::VERTEX_DATA);
+    if (!vertexData.empty())
     {
-        shader.addLine("in VertexData", false);
-        shader.beginScope(Shader::Brackets::BRACES);
-        for (const Shader::Variable* input : vertexDataBlock.variableOrder)
-        {
-            const string& type = _syntax->getTypeName(input->type);
-            shader.addLine(type + " " + input->name);
-        }
-        shader.endScope(false, false);
-        shader.addStr(" " + vertexDataBlock.instance + SEMICOLON_NEWLINE);
-        shader.newLine();
+        emitLine("in " + vertexData.getName(), stage, false);
+        emitScopeBegin(stage, ShaderStage::Brackets::BRACES);
+        emitVariableDeclarations(vertexData, EMPTY_STRING, SEMICOLON, context, stage, false);
+        emitScopeEnd(stage, false, false);
+        emitString(" " + vertexData.getInstance() + SEMICOLON, stage);
+        emitLineBreak(stage);
+        emitLineBreak(stage);
     }
 
     // Add the pixel shader output. This needs to be a vec4 for rendering
     // and upstream connection will be converted to vec4 if needed in emitFinalOutput()
-    shader.addComment("Data output by the pixel shader");
-    const ShaderGraphOutputSocket* outputSocket = shader.getGraph()->getOutputSocket();
-    shader.addLine("out vec4 " + outputSocket->variable);
-    shader.newLine();
+    emitComment("Pixel shader outputs", stage);
+    const VariableBlock& outputs = stage.getOutputBlock(HW::PIXEL_OUTPUTS);
+    emitVariableDeclarations(outputs, _syntax->getOutputQualifier(), SEMICOLON, context, stage, false);
+    emitLineBreak(stage);
 
     // Emit common math functions
-    shader.addInclude("pbrlib/" + GlslShaderGenerator::LANGUAGE + "/lib/mx_math.glsl", *this);
-    shader.newLine();
+    emitInclude("pbrlib/" + GlslShaderGenerator::LANGUAGE + "/lib/mx_math.glsl", context, stage);
+    emitLineBreak(stage);
 
     // Emit lighting functions
     if (lighting)
```

Comment:
I could, if we think it would be useful for overriding this in a derived class... However overridability in not the main reason for why I wrapped all the Stage methods into ShaderGenerator methods. The main reason is that I wanted to hide all the source code editing methods on the Stage class. They are protected and only the ShaderGenerator class is allowed to call them (being a friend). This was done to minimize the public interface on the Stage and Shader class. Since they are the product of shader generation it doesn't makes sense to call source code editing methods on them after generation is completed. I wanted to move all source code editing operations to the ShaderGenerator base class and its derived classes. Since only the base class is a friend we need wrappers that any derived class can access.
And as a bonus it makes it possible to override each of these if any customization is needed by derived classes.